### PR TITLE
bugfix: use etcd:watchcancel to reduce connections in etcdv3

### LINKED ISSUE #2295: request help: Too many connections when using etcdv3
### Issue description
using `etcd v3` find too many connections.
I deployed an apisix and started 4 workers, Up to 300+ connections

image

### Environment

* apisix version : 7beafd345a96b5fc14d9230731f284429ad7168e
* OS: Centos Linux release 7.4.1708 (Core)
image

### PR DESCRIPTION
**Pull request by Yiyiyimu**
### What this PR does / why we need it:
fix #2295 

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [ ] Have you added corresponding test cases?
* [ ] Have you modified the corresponding document?
* [x] Is this PR backward compatible?

### CODE REVIEW

File: apisix/core/config_etcd.lua
Comment (membphis) [inline apisix/core/config_etcd.lua:128]: need test case.

we can use this way: write a log and confirm it in test case

File: apisix/core/config_etcd.lua
Comment (membphis) [inline apisix/core/config_etcd.lua:256]: why we delete this code block? it is useful for etcd `v2`. please confirm this

Comment (Yiyiyimu) [reply apisix/core/config_etcd.lua:256]: Because etcd v3 does not have 1000 event history limit anymore, it could support unlimited history.

File: apisix/core/config_etcd.lua
Comment (Yiyiyimu) [inline apisix/core/config_etcd.lua:181]: Currently I still set `dir = true` in [etcd protocol conversion](https://github.com/apache/apisix/blob/9ac02f66b14fa3c1b5a8f9341a82b664ceda17fb/apisix/core/etcd.lua#L61) for consistency between protocols. But I noticed there is only one test case ([global.t TEST 3](https://github.com/apache/apisix/blob/9ac02f66b14fa3c1b5a8f9341a82b664ceda17fb/t/admin/global-rules.t#L125-L127)) needs this parameter, so maybe we could remove it

File: apisix/core/config_etcd.lua
Comment (Yiyiyimu) [reply apisix/core/config_etcd.lua:256]: But we also need to consider history compaction, or the accumulation of history would affect performance in a long term. Reference: https://etcd.io/docs/v3.4.0/op-guide/maintenance/#history-compaction

File: apisix/core/config_etcd.lua
Comment (Yiyiyimu) [reply apisix/core/config_etcd.lua]: thx, test added

File: apisix/core/config_etcd.lua
Comment (nic-chen) [reply apisix/core/config_etcd.lua:256]: @membphis  I think we don't need to consider compatibility with v2

Comment (membphis) [reply apisix/core/config_etcd.lua:256]: got it. many thanks for explain

**Comment by Yiyiyimu** [comment]
Hi @membphis is there something else we should do before merging

**Comment by Yiyiyimu** [comment]
ping @moonming it seems good to go

**Review by idbeta — APPROVED** [review]
I passed in my local environment

File: apisix/core/etcd.lua
Comment (moonming) [inline apisix/core/etcd.lua:39]: Where will it be used?

Comment (Yiyiyimu) [reply apisix/core/etcd.lua:39]: In lua-resty-etcd. If api_prefix is not set, it would check the version of etcd to find the correct api_prefix to use in uri. Since we support etcd v3.4+ only, we could directly set it to avoid useless version check each time.

**Review by moonming — APPROVED** [review]

**Review by membphis — APPROVED** [review]

File: apisix/core/config_etcd.lua
Comment (membphis) [inline apisix/core/config_etcd.lua:129]: `if res_cancel == 1 then` is enough

**Review by membphis — COMMENTED** [review]
code style

File: apisix/core/config_etcd.lua
Comment (Yiyiyimu) [reply apisix/core/config_etcd.lua:129]: Thx! Should I create a PR for it or leave it to the next PR

Comment (moonming) [reply apisix/core/config_etcd.lua:129]: sure,welcome

**Pull request merged**
